dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java

```diff
@@ -210,28 +211,31 @@ public void close() {
 
         @Override
         public T deserialize(Response response) {
-            try {
-                if (errorDecoder.isError(response)) {
+            if (errorDecoder.isError(response)) {
+                try {
                     throw errorDecoder.decode(response);
-                } else if (response.code() == 204) {
-                    if (!isOptionalType) {
-                        throw new SafeRuntimeException(
-                                "Unable to deserialize non-optional response type from 204", SafeArg.of("type", token));
-                    }
+                } finally {
+                    response.close();
+                }
+            } else if (response.code() == 204) {
+                response.close();
+                if (!isOptionalType) {
+                    throw new SafeRuntimeException(
+                            "Unable to deserialize non-optional response type from 204", SafeArg.of("type", token));
+                } else {
                     return TypeMarkers.getEmptyOptional(token);
                 }
+            }
 
-                Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
-                if (!contentType.isPresent()) {
-                    throw new SafeIllegalArgumentException(
-                            "Response is missing Content-Type header",
-                            SafeArg.of("received", response.headers().keySet()));
-                }
-                EncodingDeserializerContainer<T> container = getResponseDeserializer(contentType.get());
-                return container.deserializer.deserialize(response.body());
-            } finally {
+            Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+            if (!contentType.isPresent()) {
                 response.close();
+                throw new SafeIllegalArgumentException(
+                        "Response is missing Content-Type header",
+                        SafeArg.of("received", response.headers().keySet()));
             }
+            Encoding.Deserializer<T> deserializer = getResponseDeserializer(contentType.get());
+            return deserializer.deserialize(response.body(), response);
         }
```

Comment:
this is really the key change in this PR - previously the finally block would sweep up _all_ response instances and call close (which was bad for the binary cases), whereas now we rely on each individual deserializer to do the right thing, and this lets us put some special logic in the binary deserializers.